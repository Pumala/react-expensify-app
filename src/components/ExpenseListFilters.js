import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused });
    };
    onDatesChange = ({ startDate, endDate}) => {
        
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);        
    };
    onSortByChange = (e) => {
        const value = e.target.value;

        if (value == 'date') {
            this.props.sortByDate();
        } else {
            this.props.sortByAmount();
        }
    };
    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextFilterChange} />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}>
                    <option>Select</option>        
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>            
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={false}
                    numberOfMonths={1}
                    isOutsideRange={() => false }
                />
            </div>
        );
    };
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);