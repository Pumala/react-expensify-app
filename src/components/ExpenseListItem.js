import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
    <li>
        <p>ID: {id}</p>
        <Link to={`/edit/${id}`}>
            <p>Description: {description}</p>
        </Link>
        <p>Note: {note}</p>
        <p>Amount: {amount}</p>
        <p>Created At: {createdAt}</p>
        <button>Expense Item Page</button>
    </li>
);

export default ExpenseListItem;