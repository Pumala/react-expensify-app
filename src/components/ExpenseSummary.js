import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount > 1 ? 'expenses' : 'expense';
    const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div>
            <p>Expenses Summary El</p>
            {expensesCount > 0 && <p>
                You are viewing {expensesCount} {expenseWord} totaling {formattedExpenseTotal}.
            </p>}
            {expensesCount < 1 && <p>There are no expenses at this time.</p>}
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);