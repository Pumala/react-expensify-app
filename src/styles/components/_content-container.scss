.content-container {
    max-width: 80rem;
    margin: auto;
    padding: 0 $m-size;
    align-items: center;
}