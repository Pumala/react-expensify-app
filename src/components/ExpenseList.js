import React from 'react';
// called connect because it connects your component to the redux store
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    return (
        <div>
            <h1>Expense List</h1>
            <ul>
            {
                props.expenses.length === 0 ? (
                    <li>No expenses found.</li>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
        
                        // expense={expense} or {...expense}
                        // similar, not the same
                    })
                )
            }
            </ul>
        </div>
    )
};

// this function lets us determine what info we want from the store that 
    // we want our component to access
const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

// export HOC
export default connect(mapStateToProps)(ExpenseList); 



// Alt below
// HOC
// const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

// export default ConnectedExpenseList;