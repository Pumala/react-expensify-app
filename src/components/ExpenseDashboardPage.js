import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = (props) => {
    return (
        <div>
            <p>This is from Dashboard Component.</p>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    );
}

export default ExpenseDashboardPage;