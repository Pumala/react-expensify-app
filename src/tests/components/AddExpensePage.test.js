import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let addExpenseSpy 
let history
let wrapper;

beforeEach(() => {
    addExpenseSpy = jest.fn();
    history = {
        push: jest.fn()
    };

    wrapper = shallow(<AddExpensePage addExpense={addExpenseSpy} history={history} />);
});

test('should render AddExpensePage correctly', () => {

    expect(wrapper).toMatchSnapshot();

}); 

test('should handle obSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[1]);
});