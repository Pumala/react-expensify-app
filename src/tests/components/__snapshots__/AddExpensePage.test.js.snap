// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    This is from Add Expense Component.
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
